hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/Compactor.java

```diff
@@ -550,22 +551,63 @@ protected InternalScanner createScanner(HStore store, ScanInfo scanInfo,
         dropDeletesFromRow, dropDeletesToRow);
   }
 
-  public List<Path> getCompactionTargets() {
-    T writer = this.writer;
-    if (writer == null) {
-      return Collections.emptyList();
+  /**
+   * Return the progress for a given compaction request.
+   * @param request the compaction request
+   */
+  public CompactionProgress getProgress(CompactionRequestImpl request) {
+    return progressMap.get(request);
+  }
+
+  /**
+   * Return the aggregate progress for all currently active compactions.
+   */
+  public CompactionProgress getProgress() {
+    synchronized (progressMap) {
+      long totalCompactingKVs = 0;
+      long currentCompactedKVs = 0;
+      long totalCompactedSize = 0;
+      for (CompactionProgress progress: progressMap.values()) {
+        totalCompactingKVs += progress.totalCompactingKVs;
+        currentCompactedKVs += progress.currentCompactedKVs;
+        totalCompactedSize += progress.totalCompactedSize;
+      }
+      CompactionProgress result = new CompactionProgress(totalCompactingKVs);
+      result.currentCompactedKVs = currentCompactedKVs;
+      result.totalCompactedSize = totalCompactedSize;
+      return result;
     }
-    if (writer instanceof StoreFileWriter) {
-      return Arrays.asList(((StoreFileWriter) writer).getPath());
+  }
+
+  public boolean isCompacting() {
+    return !progressMap.isEmpty();
+  }
+
+  /**
+   * Return the list of target files for all currently active compactions.
+   */
+  public List<Path> getCompactionTargets() {
+    // Build a list of all the compaction targets for all active writers
+    List<Path> targets = new ArrayList<>();
+    synchronized (writerMap) {
+      for (T writer: writerMap.values()) {
+        if (writer instanceof StoreFileWriter) {
+          targets.add(((StoreFileWriter) writer).getPath());
+        } else {
+          ((AbstractMultiFileWriter) writer).writers().stream()
+            .forEach(sfw -> targets.add(sfw.getPath()));
+        }
+      }
     }
-    return ((AbstractMultiFileWriter) writer).writers().stream().map(sfw -> sfw.getPath())
-      .collect(Collectors.toList());
+    return targets;
   }
 
   /**
-   * Reset the Writer when the new storefiles were successfully added
+   * Complete the compaction after the new storefiles are successfully added.
    */
-  public void resetWriter(){
-    writer = null;
+  public void completeCompaction(CompactionRequestImpl request) {
+    writerMap.remove(request);
+    progressMap.remove(request);
```

Comment:
I'm assuming that you're generally OK having these two synchronized data structures updated not under the same lock? Given the current use of `writerMap` and `progressMap`, nothing is jumping out at me as bad as you currently have it. Just thinking out loud.

Reply:
Do we care if either of the `remove()` calls returns `null` (i.e. the map doesn't have the mapping in it)?